src/utils/settings.rb

```diff
@@ -0,0 +1,920 @@
+        # of errors printed by the compiler.
+        prog = StringIO.new
+        prog << "int main() {\n"
+        ii = 0
+        members = Hash.new
+        pending.each do |member, group|
+            var = "var_#{ii}"
+            members[ii] = member
+            ii += 1
+            gt = group["type"]
+            mf = member["field"]
+            prog << "#{gt} #{var}; #{var}.#{mf}.__type_detect_;\n"
+        end
+        prog << "return 0;\n"
+        prog << "};\n"
+        stderr = compile_test_file(prog)
+        stderr.scan(/var_(\d+).*?', which is of non-class type '(.*)'/).each do |m|
+            member = members[m[0].to_i]
+            case m[1]
+            when "int8_t {aka signed char}"
+                typ = "int8_t"
+            when "uint8_t {aka unsigned char}"
+                typ = "uint8_t"
+            when "int16_t {aka short int}"
+                typ = "int16_t"
+            when "int32_t {aka long int}"
+                typ = "int32_t"
+            when "uint16_t {aka short unsigned int}"
+                typ = "uint16_t"
+            when "uint32_t {aka long unsigned int}"
+                typ = "uint32_t"
+            when "float"
+                typ = "float"
+            else
+                raise "Unknown type #{m[1]} when resolving type for setting #{member["name"]}"
+            end
+            dputs "#{member["name"]} type is #{typ}"
+            member["type"] = typ
+        end
+        # Make sure all types have been resolved
+        foreach_enabled_member do |group, member|
+            if !member["type"]
+                raise "Could not resolve type for member #{member["name"]} in group #{group["name"]}"
+            end
+        end
+    end
+
+    def initialize_name_encoder
+        names = []
+        foreach_enabled_member do |group, member|
+            names << member["name"]
+        end
+        best = nil
+        (3..7).each do |v|
+            enc = NameEncoder.new(names, v)
+            if best == nil || best.estimated_size(@count) > enc.estimated_size(@count)
+                best = enc
+            end
+        end
+        dputs "Using name encoder with max_length = #{best.max_length}"
+        @name_encoder = best
+    end
+
+    def initialize_value_encoder
+        values = []
+        constants = []
+        add_value = -> (v) {
+            v = v || 0
+            if v.is_number_kind? || (v.class == String && v.is_number?)
+                values << v.to_i
+            else
+                constants << v
+            end
+        }
+        foreach_enabled_member do |group, member|
+            add_value.call(member["min"])
+            add_value.call(member["max"])
+        end
+        constantValues = resolve_constants(constants)
+        # Count values used by constants
+        constants.each do |c|
+            values << constantValues[c]
+        end
+        @value_encoder = ValueEncoder.new(values, constantValues)
+    end
+
+    def resolve_constants(constants)
```

Comment:
Same as above ... In this case it may be possible to generate initialized `int64_t` array and parse it's contents from generated object file